tests/test_ingestion.py

```diff
@@ -0,0 +1,183 @@
+        """Return multiple ways of querying calibration repositories.
+
+        Returns
+        -------
+        butler : `lsst.daf.persistence.Butler`
+            A butler that should all be capable of finding ingested science data.
+        """
+        return dafPersist.Butler(inputs={'root': self._repo, 'mode': 'r'})
+
+    def _calibButler(self):
+        """Return multiple ways of querying calibration repositories.
+
+        Returns
+        -------
+        butlers : `lsst.daf.persistence.Butler`
+            A butler that should all be capable of finding ingested calibration data.
+        """
+        return dafPersist.Butler(inputs={
+            'root': self._repo,
+            'mode': 'r',
+            'mapperArgs': {'calibRoot': self._calibRepo}})
+
+    def tearDown(self):
+        shutil.rmtree(self._repo, ignore_errors=True)
+
+    def testDataIngest(self):
+        """Test that ingesting a science image adds it to a repository.
+        """
+        files = [os.path.join(IngestionTestSuite.testData,
+                              'rawData',
+                              '2013-09-01',
+                              'z',
+                              'decam0229388.fits.fz')]
+        ingestion._doIngest(self._repo, IngestionTestSuite.testApVerifyData, files)
+
+        butler = self._rawButler()
+        self.assertTrue(butler.datasetExists('raw', dataId=IngestionTestSuite.rawDataId))
+
+    def testCalibIngest(self):
+        """Test that ingesting calibrations adds them to a repository.
+        """
+        files = [os.path.join(IngestionTestSuite.testData,
+                              'rawData',
+                              'cpCalib',
+                              'masterCal',
+                              calibFile) for calibFile in
+                 [os.path.join('fci.fits'),
+                  os.path.join('zci.fits')]
+                 ]
+
+        ingestion._flatBiasIngest(self._repo, self._calibRepo, files)
+
+        butler = self._calibButler()
+        self.assertTrue(butler.datasetExists('cpBias', dataId=IngestionTestSuite.calibDataId))
+        self.assertTrue(butler.datasetExists('cpFlat', dataId=IngestionTestSuite.calibDataId))
+
+    def testDefectIngest(self):
+        """Test that ingesting defects adds them to a repository.
+        """
+        files = [os.path.join(IngestionTestSuite.testApVerifyData, 'defects'),
+                 os.path.basename(IngestionTestSuite.defectDataId['path'])]
+
+        ingestion._defectIngest(self._repo, self._calibRepo, files)
+
+        butler = self._calibButler()
+        self.assertTrue(butler.datasetExists('defects', dataId=IngestionTestSuite.defectDataId))
+
+    @unittest.skip("Ingestion functions cannot handle empty file lists, see DM-13835")
+    @unittest.skip("Dataset enumeration requires specific data keys for date, filter, etc., see DM-12762")
+    def testNoFileIngest(self):
+        """Test that attempts to ingest nothing do nothing.
+        """
```

Comment:
I'm not convinced that loudly crashing is a bad thing if a user tries to ingest nothing, because I can't think of a use case where I would intentionally want to ingest nothing. Am I missing something?

Reply:
I can't think of a specific use case either, but crashing over something like this is extremely unusual/surprising behavior. For example, if you try to iterate over an empty list, you expect the loop to simply do nothing instead of punishing you for not writing special-case code that would have amounted to the same thing.

Also, having a low-level function exit because of invalid inputs is extremely bad practice. If we want to ban empty file lists, the function should throw an exception so that the *program* can decide whether or not an exit is warranted (and have a chance to clean up properly either way).